pydantic/_internal/_generate_schema.py

```diff
@@ -312,22 +312,39 @@ def _generate_parameter_schema(
             parameter_schema['alias'] = field.alias
         return parameter_schema
 
-    def _generic_collection_schema(self, type_: Any) -> core_schema.CoreSchema:
+    def _generic_collection_schema(
+        self, type_: type[typing.List[Any]] | type[typing.Set[Any]] | type[typing.FrozenSet[Any]]
+    ) -> core_schema.CoreSchema:
         """
         Generate schema for List, Set etc. - where the schema includes `items_schema`
 
         e.g. `list[int]`.
         """
-        try:
-            name = type_.__name__
-        except AttributeError:
-            name = get_origin(type_).__name__  # type: ignore[union-attr]
-
-        return {  # type: ignore[misc,return-value]
-            'type': name.lower(),
+        # For the sake of schema generation, if `type_` is a proper subclass of one of the generic collections,
+        # we want to determine the specific collection type to use in the schema; this is stored in schema_type
+        schema_type: Any = type_
+        if not hasattr(schema_type, '__name__'):
+            schema_type = get_origin(type_)
+
+        is_proper_subclass = False
+        if schema_type not in {list, set, frozenset} and hasattr(schema_type, '__bases__'):
+            for base in schema_type.__bases__:
+                if base in {list, set, frozenset}:
+                    schema_type = base
+                    is_proper_subclass = True
+                    break
+
+        schema: core_schema.CoreSchema = {  # type: ignore[misc,assignment]
+            'type': schema_type.__name__.lower(),
             'items_schema': self.generate_schema(get_first_arg(type_)),
         }
 
+        if is_proper_subclass:
+            # Ensure the validated value is converted back to the specific subclass type
+            return core_schema.function_after_schema(schema, lambda __input_value, **kwargs: type_(__input_value))
```

Comment:
@samuelcolvin Is this along the lines you had in mind? I took the idea of the `function_after_schema` call from `deque_schema`, and otherwise just did modifications to the `_generic_collection_schema` function.

It might be possible to pull the handling of proper subclasses into a separate function if you don't want to muddle the `_generic_collection_schema` method as much; let me know what you think.

Reply:
I started trying to explain what I thought through a comment, and decided it was easier to contribute directly to the PR.

I hope you don't mind.

I think this approach is easier, we re-use information we already had outside the function. Let me know what you think.

No idea if it works in all cases, but it works in more cases than what we had before this PR :-).